File: packages/kotlinModule/src/Update.kt
/**
 * Created by ryota on 2017/03/05.
 */
fun appUpdate(msg: Msg, model: Model): Model {
    return when (msg) {
        is Msg.Init -> model
        is Msg.Update -> model.copy(num = msg.num)
        is Msg.AppChildMsg -> model.copy(childModel = child.update(model.childModel, msg.childMsg))
        else -> model
    }
}

val update: (msg: Msg, model: Model) -> Model = { msg, model -> appUpdate(msg, model) }

File: packages/scalaModule/src/main/scala/application/Update.scala
package application

import base.AbstractUpdate

object Update extends AbstractUpdate[Msg, Model]  {
  def apply(msg: Msg, model: Model) = msg match {
    case Init => model
    case SetText(text) => model.copy(text = text)
    case ChildMsg(childMsg) => model.copy(child = child.Update.apply(childMsg, model.child))
  }
}